File: node/lib/revision.mjs
// revision — read document revision history (read-only).
import { apiPost, extractId } from "./api.mjs";
import { renderResponse, printJson } from "./utils.mjs";

function summRevision(r, previewLen = 120) {
  if (!r || typeof r !== "object") return r;
  const by = r.createdBy || {};
  const out = {
    id: r.id,
    documentId: r.documentId,
    title: r.title,
    createdBy: { id: by.id, name: by.name },
    createdAt: r.createdAt,
  };
  if (previewLen && previewLen > 0) {
    const t = r.text || "";
    out.textPreview = t.length > previewLen ? t.slice(0, previewLen) + "…" : t;
  }
  return out;
}

export async function handleRevision(action, f) {
  switch (action) {
    case "list": {
      if (!f.documentId) { printJson({ ok: false, error: "需要 --document-id" }); process.exit(1); }
      const p = { documentId: extractId(f.documentId, "doc") };
      if (f.limit) p.limit = parseInt(f.limit, 10);
      if (f.offset) p.offset = parseInt(f.offset, 10);
      renderResponse(await apiPost("revisions.list", p), summRevision, f.full, f);
      return;
    }
    case "info": {
      if (!f.id) { printJson({ ok: false, error: "需要 --id" }); process.exit(1); }
      renderResponse(await apiPost("revisions.info", { id: f.id }), summRevision, f.full, f);
      return;
    }
    default:
      printJson({ ok: false, error: `未知 revision 动作：${action}（可用：list, info）` });
      process.exit(2);
  }
}

File: python/scripts/revision.py
"""revision subcommand — read document revision history (read-only)."""
import sys
from .api import api_post, extract_id
from .utils import render_response, print_json


def _summ_revision(r, preview_len=120):
    if not isinstance(r, dict):
        return r
    by = r.get("createdBy") or {}
    out = {
        "id": r.get("id"),
        "documentId": r.get("documentId"),
        "title": r.get("title"),
        "createdBy": {"id": by.get("id"), "name": by.get("name")},
        "createdAt": r.get("createdAt"),
    }
    if preview_len and preview_len > 0:
        t = r.get("text") or ""
        out["textPreview"] = t[:preview_len] + "…" if len(t) > preview_len else t
    return out


def handle_revision(args):
    if args.action == "list":
        if not args.document_id:
            print_json({"ok": False, "error": "需要 --document-id"}); sys.exit(1)
        payload = {"documentId": extract_id(args.document_id, "doc")}
        if getattr(args, "limit", None): payload["limit"] = args.limit
        if getattr(args, "offset", None): payload["offset"] = args.offset
        res = api_post("revisions.list", payload)
        render_response(res, summarizer=_summ_revision, full=args.full)

    elif args.action == "info":
        if not args.id:
            print_json({"ok": False, "error": "需要 --id"}); sys.exit(1)
        res = api_post("revisions.info", {"id": args.id})
        render_response(res, summarizer=_summ_revision, full=args.full)


def setup_revision_parser(subparsers):
    p = subparsers.add_parser("revision", help="文档历史版本（只读）")
    p.add_argument("action", choices=["list", "info"])
    p.add_argument("--id", help="revision ID")
    p.add_argument("--document-id", help="文档 ID 或 URL")
    p.add_argument("--limit", type=int)
    p.add_argument("--offset", type=int)
    p.add_argument("--full", action="store_true")
    p.set_defaults(func=handle_revision)
